[InstCombine] If inst in unreachable refers to an inst change it to poison (#65107)

ParkHanbum opened the pull request:
Instructions in unreachable basic blocks are removed, but terminators
are not. In this case, even instructions that are only referenced by
a terminator, such as a return instruction, cannot be processed
properly.

This patch changes the operand of a return instruction in an
unreachable basic block to poison if it refers to the instruction,
allowing the instruction to be properly processed.

nikic commented:
As I wrote in https://github.com/llvm/llvm-project/issues/65107#issuecomment-1890391251, I don't think this is the correct way to fix the issue.

ParkHanbum commented:
> As I wrote in [#65107 (comment)](https://github.com/llvm/llvm-project/issues/65107#issuecomment-1890391251), I don't think this is the correct way to fix the issue.

I read your comment but I didn't know what that mean. what we do incorrectly exactly?

nikic commented:
@ParkHanbum My suggestion is to fold
```llvm
define i1 @f(i16 %0) {
  br label %BB1

BB1:                                              ; preds = %BB1, %BB
  %C = icmp ult i16 256, %0
  br i1 true, label %BB1, label %BB2

BB2:                                              ; preds = %BB1
  ret i1 %C
}
```
to
```llvm
define i1 @f(i16 %0) {
  br label %BB1

BB1:                                              ; preds = %BB1, %BB
  %C = icmp ult i16 256, %0
  br i1 true, label %BB1, label %BB2

BB2:                                              ; preds = %BB1
  ret i1 poison
}
```
at which point `%C` becomes dead and can be removed, without having it to sink into BB2 first.

This is a more general solution, which allows us to handle additional cases. For example, cases where we wouldn't sink the instruction due to complex CFG, and cases where the instruction is used both in the dead block and somewhere else, in which case removing the use in the dead block may enable a one-use fold.

ParkHanbum commented:
@nikic sounds reasonable. I'll change it like that!

ParkHanbum committed:
[InstCombine] If Inst in unreachable refers to Inst change it to poison (#65107)

Instructions in unreachable basic blocks are removed, but terminators
are not. In this case, even instructions that are only referenced by
a terminator, such as a return instruction, cannot be processed
properly.

This patch changes the operand of terminator instruction in an
unreachable basic block to poison if it refers to the instruction,
allowing the instruction to be properly processed.

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -2779,6 +2779,11 @@ llvm::removeAllNonTerminatorAndEHPadInstructions(BasicBlock *BB) {
+
+  if (isa<ReturnInst>(EndInst) && EndInst->getNumOperands() > 0 &&
+      isa<Instruction>(EndInst->getOperand(0)))
+    EndInst->setOperand(0, PoisonValue::get(EndInst->getOperand(0)->getType()));
```
We shouldn't hardcode this to `ReturnInst`. We should replace any operand in the terminator. E.g. if this is a `br`, we should also set the condition to poison. (Skip the replacement for token types though.)

You should also do a similar change in `InstCombinerImpl::handleUnreachableFrom`.

ParkHanbum replied:
@nikic got it. 
//‘ret’, ‘br’, 
// ‘switch’, ‘indirectbr’, ‘invoke’, ‘callbr’ ‘resume’, 
//‘catchswitch’, ‘catchret’, ‘cleanupret’, and ‘unreachable’.

I'll implement codes to handle above terminators.
except `CatchSwitch`. it is element of EHPad can be skipped.

nikic replied:
The handling should be generic, not for specific terminator types.

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -2770,15 +2770,31 @@ bool llvm::replaceAllDbgUsesWith(Instruction &From, Value &To,
+  // RemoveDIs: erase debug-info on this instruction manually.
+  I->dropDbgValues();
+  unsigned Opcode = I->getOpcode();
+  if (Opcode >= Instruction::Ret && Opcode <= Instruction::Invoke &&
```
Why do we need this opcode check?

ParkHanbum replied:
I thought that only the above commands could be processed among the terminators. This is just my opinion, so please let me know if I'm wrong. 

this is my sample IRs.
```
icatch.dispatch:
  %tmp1 = catchswitch within none [label %icatch] unwind to caller
icatch:
  %tmp2 = catchpad within %tmp1 [ptr null, i32 64, ptr null]
  catchret from %tmp2 to label %exit
unreach.cleanupret:
  %cl = cleanuppad within none []
  cleanupret from %cl unwind to caller

unreach.indirectbr:
  indirectbr ptr %gep, [label %unreach.switch]

unreach.callbr:
  callbr void asm "", "!i"()
  to label %dummy [label %toexit]
```

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -2770,15 +2770,31 @@ bool llvm::replaceAllDbgUsesWith(Instruction &From, Value &To,
+  unsigned Opcode = I->getOpcode();
+  if (Opcode >= Instruction::Ret && Opcode <= Instruction::Invoke &&
+      I->getNumOperands() > 0) {
+    unsigned OpNum = isa<InvokeInst>(I) ? I->getNumOperands() - 1 : 0;
```
You should loop over `I->operands()` here, not only update a single operand.

ParkHanbum replied:
you mean set the poison to all of operand?

ParkHanbum replied:
it has be done!

nikic commented on llvm/test/Transforms/InstCombine/sink_to_unreachable.ll:
```diff
@@ -157,3 +157,141 @@ bb3:
+  br i1 true, label %loop, label %unreach
+
+unreach:
+  invoke void %gep(i1 false)
```
Also add a call operand that can be replaced.

ParkHanbum replied:
I missed it! I'll update it ASAP

ParkHanbum replied:
but call instruction is not belong to terminator, isn't it?

nikic replied:
I meant an argument to the invoke. You currently have i1 false, but you can also pass an instruction to it, which could be replaced with poison.

ParkHanbum replied:
it has be done!

dtcxzyw commented:
Could you have a look at https://github.com/dtcxzyw/llvm-opt-benchmark/pull/226#discussion_r1487162964?
It is wired to replace an argument in a reachable invocation with a poison.

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -2770,15 +2770,32 @@ bool llvm::replaceAllDbgUsesWith(Instruction &From, Value &To,
+  unsigned Opcode = I->getOpcode();
+  if (Opcode >= Instruction::Ret && Opcode <= Instruction::Invoke &&
+      I->getNumOperands() > 0) {
+    for (unsigned i = 0; i < I->getNumOperands(); i++) {
```
```suggestion
    for (Use &U : I->operands())
```
and then `U.set()`.

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3100,9 +3100,8 @@ void InstCombinerImpl::handleUnreachableFrom(
-  // terminator too.
-  BB->getTerminator()->dropDbgValues();
+  // if terminator refer instruction, transform it into poison.
+  handleUnreachableTerminator(BB->getTerminator());
```
```suggestion
  MadeIRChange |= handleUnreachableTerminator(BB->getTerminator());
```
and return whether a change was made from that function.

ParkHanbum commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -2762,15 +2762,30 @@ bool llvm::replaceAllDbgUsesWith(Instruction &From, Value &To,
+  I->dropDbgValues();
+  for (Use &U : I->operands()) {
+    Value *Op = U.get();
+    if (isa<Instruction>(Op) && !Op->getType()->isTokenTy()) {
```
ok. I have a question. when Op is Poison it can pass `isa<Instruction>(Op)` condition?

nikic replied:
Nope :) You are right, checking poison is unnecessary.

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3231,9 +3231,8 @@ void InstCombinerImpl::handleUnreachableFrom(
-  // RemoveDIs: to match behaviour in dbg.value mode, drop debug-info on
-  // terminator too.
-  BB->getTerminator()->dropDbgValues();
+  if (!DT.isReachableFromEntry(BB))
```
Your previous version here was correct. handleUnreachableFrom already guarantees that the code is unreachable. Checking isReachableFromEntry does not work as desired inside InstCombine (it does not make use of reachability knowledge private to InstCombine).

nikic commented on llvm/include/llvm/Transforms/Utils/Local.h:
```diff
@@ -357,6 +357,13 @@ Value *salvageDebugInfoImpl(Instruction &I, uint64_t CurrentLocOps,
+/// If a terminator in an unreachable basic block has an operand of type
+/// Instruction, transform it into poison. Return true if any operands
+/// are changed to poison. Original Values prior to being changed to poison
+/// be returned to PoisonedValues.
```
```suggestion
/// are returned in PoisonedValues.
```

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3231,9 +3231,12 @@ void InstCombinerImpl::handleUnreachableFrom(
-  // RemoveDIs: to match behaviour in dbg.value mode, drop debug-info on
-  // terminator too.
-  BB->getTerminator()->dropDbgValues();
+  SmallVector<Value *, 4> Changed;
```
```suggestion
  SmallVector<Value *> Changed;
```

nikic reviewed (approved):
LGTM

dtcxzyw commented on llvm/include/llvm/Transforms/Utils/Local.h:
```diff
@@ -357,6 +357,13 @@ Value *salvageDebugInfoImpl(Instruction &I, uint64_t CurrentLocOps,
+/// If a terminator in an unreachable basic block has an operand of type
+/// Instruction, transform it into poison. Return true if any operands
+/// are changed to poison. Original Values prior to being changed to poison
+/// are returned to PoisonedValues.
```
```suggestion
/// are returned in \p PoisonedValues.
```

nikic committed:
Update llvm/include/llvm/Transforms/Utils/Local.h

Co-authored-by: Yingwei Zheng <dtcxzyw@qq.com>

aaronpuchert commented:
As this fixes a crash, do you think we should port it back to LLVM 18? We're [running into this](https://bugzilla.opensuse.org/show_bug.cgi?id=1221664) with KDE 6 on aarch64.

CC @tstellar.

nikic commented:
@aaronpuchert No, this change shouldn't be backported -- it's an optimization improvement, not a bug fix. The error is a development tool, not intended to surface to users.

You probably need to backport this mesa patch: https://gitlab.freedesktop.org/mesa/mesa/-/commit/99f0449987bec1f82cd42a06f40bb4a863a37792 We ran into a similar openQA failure in Fedora, and this is the fix for it.

aaronpuchert commented:
Thanks, I was just confused by the crash tag on #65107. Will forward your suggestion.
